### PR TITLE
fixes dotnet/templating#2919: NuGet API refactoring

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
fixes dotnet/templating#2919: NuGet API refactoring
based on feedback in https://github.com/dotnet/templating/pull/2878

### Solution
- changed order of source processing
- parallelize processing for multiple sources

### Checks:
- [ ] Added unit tests
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (DavidKarlas) [inline src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:303]: Is this commented code intentional?

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (DavidKarlas) [inline src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:385]: Is this just about perf or also changes behavior? Wondering how hard would be to unit test...

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (DavidKarlas) [inline src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:178]: `includePrerelease: true`, but this method has `includePreview` parameter, should it be used?

Comment (DavidKarlas) [reply src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:178]: Nevermind, read rest of method.

**Review by DavidKarlas — APPROVED** [review]

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:351]: it doesn't change the behavior; in fact with parallelization everywhere the impact is minimum.

The idea behind it is that if the user specifies the source the package is likely to be found in that source, and not the default ones. Previously checking for the packages was sequential until the package is found, so it mattered. Now that processing is parallel it doesn't matter that much.

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (bozturkMSFT) [inline src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:212]: If no stable version is found, we are directly returning a preview version if found. However, if `includePreview` is false, we should probably return `null`.

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (bozturkMSFT) [inline src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:208]: These two queries can be merged into one, if you use `includePreview` inside the anonymous method. Then you can also get rid of the `if` check below.

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (bozturkMSFT) [inline src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:231]: Typo: match

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (bozturkMSFT) [inline src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:234]: We don't need the `internalCancelToken` here. On the second line, using `CreateLinkedTokenSource(cancellationToken)` will already give us the ability to cancel "internally" using `linkedCts.Cancel()`.

**Review by bozturkMSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:212]: this is intentionally to allow to install latest preview if no stable.

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs:208]: I'm not sure if this is possible - we need to check first if  the `latestStableVersion` is available. 
We can shuffle the statements around though to check for preview versions only if stable version is not available / not needed, but I'm not convinced for now if it makes code more readable - will try it out.

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (bozturkMSFT) [reply src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs]: I see. I believe the way nuget references work in dotnet today, unless you specifically asked for a preview, you will only be given a stable version or a "package not found" error. Wouldn't it be better to have the same experience across all dotnet? I will approve and leave this up to you.

**Review by bozturkMSFT — APPROVED** [review]

File: src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs
Comment (bozturkMSFT) [reply src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs]: Ah, you are right. My bad.

**Pull request merged**
